### PR TITLE
feat(api): add project for cyclone CRD resources

### PR DESCRIPTION
**Pull request by zhujian7**
**What this PR does / why we need it**:

- update resources CRUD
- update stage CRUD
- update workflow CRUD
- update workflowtrigger CRUD
- update workflowrun CRUD
- update logs related apis

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @your-reviewer

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/server/handler/v1alpha1/helper.go [hunk 1]
```diff
@@ -29,3 +39,133 @@ func getLogFolder(workflowrun, stage, namespace string) (string, error) {
...
+		meta.Annotations[common.AnnotationAlias] = name
+	} else {
+		meta.Annotations[common.AnnotationAlias] = alias
+	}
+
+	// add project label
+	if meta.Labels == nil {
+		meta.Labels = make(map[string]string)
+	}
+	meta.Labels[common.LabelProject] = project
```
Comment (cd1989) [inline pkg/server/handler/v1alpha1/helper.go]: How about this:

```
// Add project label
if meta.Labels == nil {
	meta.Labels = make(map[string]string)
}
meta.Labels[common.LabelProject] = project

// Get name and alias, if alias not set, use name as alias
name := meta.Name
alias := ""
if meta.Annotations != nil {
	alias = meta.Annotations[common.AnnotationAlias]
}
if alias == "" {
	alias = name
}

// Add alias annotation if not set
if meta.Annotations == nil {
	meta.Annotations = make(map[string]string)
}
meta.Annotations[common.AnnotationAlias] = alias

// If resource name set, check whether name conflict exists.
if name != "" {
	old, err := getMetadata(tenant, name)
	if err == nil {
	   return errors.NewAlreadyExists(schema.GroupResource{Group: v1alpha1.APIVersion, Resource: resource}, name)
	}
	return nil
}

// If resource name not set, generate one from alias.
name = slugify.Slugify(project+"-"+alias, false, -1)
old, err := getMetadata(tenant, name)
if err == nil {
	name = slugify.Slugify(name, true, -1)
}
meta.Name = name 

return nil

```

File: pkg/server/handler/v1alpha1/helper.go [hunk 2]
```diff
@@ -29,3 +39,133 @@ func getLogFolder(workflowrun, stage, namespace string) (string, error) {
...
+	resource, err := handler.K8sClient.CycloneV1alpha1().WorkflowTriggers(common.TenantNamespace(tenant)).Get(name, meta_v1.GetOptions{})
+	if err != nil {
+		return meta_v1.ObjectMeta{}, err
+	}
+	return resource.ObjectMeta, nil
+}
+
+// CreatePrelude is the prelude of create cyclone CRD resources.
+// It will give the resource a name if it is empty. and will add project labels for the project.
+func CreatePrelude(project, tenant string, object interface{}) error {
```
Comment (cd1989) [inline pkg/server/handler/v1alpha1/helper.go]: `ModifyResource`

**Comment by zhujian7** [comment]
@cd1989 All comments have been fixed, PTAL.

File: pkg/server/apis/v1alpha1/descriptors/workflowtrigger.go [hunk 3]
```diff
@@ -29,47 +29,72 @@ func init() {
...
+						Source: definition.Path,
+						Name:   httputil.ProjectNamePathParameterName,
+					},
+					{
+						Source: definition.Header,
+						Name:   httputil.TenantHeaderName,
+					},
+					{
+						Source:      definition.Auto,
+						Name:        "pagination",
```
Comment (supereagle) [inline pkg/server/apis/v1alpha1/descriptors/workflowtrigger.go]: This is also used in some other places, define a const for this?

File: pkg/server/apis/v1alpha1/descriptors/workflowrun.go [hunk 4]
```diff
@@ -91,62 +145,214 @@ var workflowrun = []definition.Descriptor{
...
+						Source: definition.Path,
+						Name:   httputil.WorkflowRunNamePathParameterName,
+					},
+					{
+						Source: definition.Header,
+						Name:   httputil.TenantHeaderName,
+					},
+					{
+						Source: definition.Query,
+						Name:   "stage",
```
Comment (supereagle) [inline pkg/server/apis/v1alpha1/descriptors/workflowrun.go]: Define a const for this?

**Pull request merged**